### PR TITLE
Deprecate util/dev in favor of cargo alias

### PR DESCRIPTION
**Pull request by phansch**
This means one less shell script and a bit more cross-platform support
for contributors.

If you've been using `./util/dev` before, this now becomes `cargo dev`.

The key part of this change is found in `.cargo/config` where an alias for calling the `clippy_dev` binary is defined.

changelog: none

### CODE REVIEW

**Comment by phansch** [comment]
Not sure if this will pass CI already and it will conflict with #5088 for now. 

Also not sure about what this part in `./util/dev` does, it may just be for CI?

```shell
CARGO_TARGET_DIR=$(pwd)/target/
export CARGO_TARGET_DIR
```

**Comment by flip1995** [comment]
Don't worry about #5088, it's an easy change there

The env var makes sure, that clippy_dev gets build in the same target dir as clippy itself. This way libraries can be shared between clippy and clippy_dev. I think there is a cargo flag to also set this.

File: .cargo/config [hunk 1]
```diff
@@ -1,5 +1,6 @@
 [alias]
 uitest = "test --test compile-test"
+dev = "run --package clippy_dev --bin clippy_dev --manifest-path clippy_dev/Cargo.toml --"
```
Comment (lzutao) [inline .cargo/config]: Use --target-dir flag I believe

**Comment by phansch** [comment]
Regarding the `CARGO_TARGET_DIR` value, I believe we don't need to set it anymore. What `util/dev` did, was setting it to be the same as the root project's target dir before `cd`ing into `clippy_dev`. When running `cargo dev foo` from the root project, it's already correct because we don't `cd` anymore.

**Comment by phansch** [comment]
GitHub seems to show the wrong commit order: 3036a2c3 is the second commit and contains some more cleanup refactoring for finding the project root directory.

The PR should be ready for review now

File: clippy_dev/src/fmt.rs [hunk 2]
```diff
@@ -176,22 +173,3 @@ fn rustfmt(context: &FmtContext, path: &Path) -> Result<bool, CliError> {
     }
     Ok(success)
 }
-
-fn project_root() -> Result<PathBuf, CliError> {
```
Comment (flip1995) [inline clippy_dev/src/fmt.rs]: This function is also copy pasted in 
https://github.com/rust-lang/rust-clippy/blob/bbef531518a3b4977a9c1f40808f950514b400fc/clippy_dev/src/new_lint.rs#L85-L101

File: clippy_dev/src/lib.rs [hunk 3]
```diff
@@ -318,6 +321,25 @@ where
     }
 }
 
+/// Returns the path to the Clippy project directory
+pub fn clippy_project_root() -> PathBuf {
```
Comment (flip1995) [inline clippy_dev/src/lib.rs]:
```suggestion
#[must_use]
pub fn clippy_project_root() -> PathBuf {
```

**Comment by lzutao** [comment]
> When running `cargo dev foo` from the root project, it's already correct because we don't `cd` anymore

No, it is not. When unset CARGO_TARGET_DIR env, running `cargo dev fmt` will run `clippy_dev/target/debug/clippy_dev`, not `target/debug/clippy_dev`.

**Pull request merged**